### PR TITLE
feat: lowering tk2ops -> hseriesops

### LINKED ISSUE #490: Lower quantum ops in tket2-hseries
Lower all Tk2Ops to HSeriesOps

### PR DESCRIPTION
**Pull request by ss2165**
Closes #490 

Using definitions in hqslib1 https://github.com/CQCL/pytket-quantinuum/blob/main/pytket/extensions/quantinuum/backends/hqslib1.inc

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -26,8 +26,8 @@ missing_docs = "warn"
 [workspace.dependencies]
 
 # Make sure to run `just recompile-eccs` if the hugr serialisation format changes.
-hugr = "0.12.0"
-hugr-core = "0.9.0"
+hugr = "0.12.1"
```
Comment (ss2165) [inline Cargo.toml:29]: to get lowering functions

**Comment by cqc-alec** [comment]
> I would like a second reviewer to check the `HSeriesOpBuilder::build_*` functions, in particular that the comments describe the correct lowering. perhaps @cqc-alec?

Will do.

File: tket2-hseries/src/extension/hseries.rs [hunk 2]
```diff
@@ -186,6 +195,248 @@ pub trait HSeriesOpBuilder: Dataflow {
         self.add_dataflow_op(HSeriesOp::QFree, [qb])?;
         Ok(())
     }
+
+    /// Build a hadamard gate in terms of HSeries primitives.
+    fn build_h(&mut self, qb: Wire) -> Result<Wire, BuildError>
+    where
+        Self: Sized,
```
[the remark this answers is not in the document]
Comment (ss2165) [reply tket2-hseries/src/extension/hseries.rs:202]: i just needed to relax it in pi_mul, thanks

File: tket2-hseries/src/extension/hseries/lower.rs [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+        }
+        _ => return Err(LowerBuildError::UnknownOp(op, inputs.len())), // non-exhaustive
+    };
+    Ok(b.finish_hugr_with_outputs(outputs, &REGISTRY)?)
+}
+
+/// Lower `Tk2Op` operations to `HSeriesOp` operations.
+pub fn lower_tk2_op(
+    mut hugr: impl HugrMut,
+) -> Result<Vec<hugr::Node>, Either<HugrError, hugr::algorithms::lower::LowerError>> {
```
[the remark this answers is not in the document]
Comment (ss2165) [reply tket2-hseries/src/extension/hseries/lower.rs:83]: you're right, removd

File: tket2-hseries/src/extension/hseries/lower.rs [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+            let float = b.add_atorad(*angle)?;
+            b.build_crz(*c, *t, float)?.into()
+        }
+        (Tk2Op::Toffoli, [a, b_, c]) => b.build_toffoli(*a, *b_, *c)?.into(),
+        (Tk2Op::QAlloc | Tk2Op::QFree | Tk2Op::Reset | Tk2Op::Measure, _) => {
+            unreachable!("should be covered by lower_direct")
+        }
+        _ => return Err(LowerBuildError::UnknownOp(op, inputs.len())), // non-exhaustive
+    };
+    Ok(b.finish_hugr_with_outputs(outputs, &REGISTRY)?)
```
[the remark this answers is not in the document]
Comment (ss2165) [reply tket2-hseries/src/extension/hseries/lower.rs:77]: this is only validating the replacement hugr, which we know to be valid under the
registry. So should be ok? In theory we could use a smaller registry to validate this.
Maybe we should have views in to registries that limits them to a subset of their
extensions (without copying extensions)?

File: tket2-hseries/src/extension/hseries.rs [hunk 5]
```diff
@@ -186,6 +195,248 @@ pub trait HSeriesOpBuilder: Dataflow {
...
+        let [c, t] = self.add_zz_max(c, t)?;
+        let c = self.add_rz(c, pi_minus_2)?;
+        let t = self.add_phased_x(t, pi_2, pi)?;
+        let t = self.add_rz(t, pi_minus_2)?;
+
+        Ok([c, t])
+    }
+
+    /// Build a CY gate in terms of HSeries primitives.
+    fn build_cy(&mut self, a: Wire, b: Wire) -> Result<[Wire; 2], BuildError>
```
Comment (cqc-alec) [inline tket2-hseries/src/extension/hseries.rs]: CY[0,1] = PhasedX(-0.5, 0)[1] ; ZZMax[0,1] ; Rz(0.5)[0] ; PhasedX(0.5, 0.5)[1] ; Rz(-0.5)[1]

File: tket2-hseries/src/extension/hseries.rs [hunk 6]
```diff
@@ -186,6 +195,248 @@ pub trait HSeriesOpBuilder: Dataflow {
...
+        // gate ry(theta) a
+        // {
+        //    phased_x(theta, pi/2) a;
+        // }
+        let pi_2 = pi_mul(self, 0.5);
+        self.add_phased_x(qb, theta, pi_2)
+    }
+
+    /// Build a CRZ gate in terms of HSeries primitives.
+    fn build_crz(&mut self, a: Wire, b: Wire, lambda: Wire) -> Result<[Wire; 2], BuildError>
```
Comment (cqc-alec) [inline tket2-hseries/src/extension/hseries.rs]: CRz(a)[0,1] = PhasedX(0.5, 0.5)[0] ; ZZMax[0,1] ; PhasedX(-1 + 0.5*a, -0.5)[0] ; ZZMax[0,1] ; PhasedX(0.5, 0.5)[0] ; Rz(0.5*a)[1]

**Comment by ss2165** [comment]
Have used a simpler zzphase decomposition for crz, I believe the rest are up to the mark with tket synthesis to the gate set

File: tket2-hseries/src/extension/hseries.rs [hunk 7]
```diff
@@ -158,34 +168,220 @@ pub trait HSeriesOpBuilder: Dataflow {
...
+            .add_dataflow_op(FloatOps::fneg, [lambda_2])?
+            .out_wire(0);
+
+        let [a, b] = self.add_zz_phase(a, b, lambda_minus_2)?;
+        let b = self.add_rz(b, lambda_2)?;
+        Ok([a, b])
+    }
+
+    /// Build a Toffoli (CCX) gate in terms of HSeries primitives.
+    fn build_toffoli(&mut self, a: Wire, b: Wire, c: Wire) -> Result<[Wire; 3], BuildError> {
```
Comment (cqc-alec) [inline tket2-hseries/src/extension/hseries.rs:357]: CCX = PhasedX(1,1)[2] ; ZZMax[1,2] ; PhasedX(0.25,-0.5)[2] ; ZZMax[0,2] ; PhasedX(-0.25,0)[2] ; ZZMax[1,2] ; PhasedX(-0.5,0.25)[1] ; PhasedX(0.25,0.5)[2] ; ZZMax[0,2] ; ZZMax[0,1] ; PhasedX(-0.75,0)[2] ; PhasedX(0.25, 0.25)[1] ; ZZMax[0,1] ; Rz(0.25)[0] ; PhasedX(-0.5, 0.25)[1] ; Rz(0.25)[1]

**Review by cqc-alec — COMMENTED** [review]
Feel free to ignore these comments as the number of 2-qubit gates is the same

**Review by cqc-alec — APPROVED** [review]

**Pull request merged**